### PR TITLE
allow remote torchscript call to itself

### PR DESCRIPTION
**Pull request by zhaojuanmao**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32959 jit pickling rref
* **#32990 allow remote torchscript call to itself**

right now remote torchscript call can not call to itself, this diff is to support this in the same way as how is supported when calling remote python call to itself

Differential Revision: [D19731910](https://our.internmc.facebook.com/intern/diff/D19731910/)

### CODE REVIEW

**Comment by zhaojuanmao** [comment]
lint and test failures are not relevant

File: torch/csrc/distributed/rpc/request_callback_impl.cpp [hunk 1]
```diff
@@ -119,11 +120,10 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
           stack.size());
 
       ownerRRef->setValue(std::move(stack.front()));
-      ctx.addForkOfOwner(
-          scriptRemoteCall.retRRefId(), scriptRemoteCall.retForkId());
-      return wrap(
-          RemoteRet(scriptRemoteCall.retRRefId(), scriptRemoteCall.retForkId())
-              .toMessage());
+      if (rrefId != forkId) {
```
Comment (xush6528) [inline torch/csrc/distributed/rpc/request_callback_impl.cpp]: I feel we need some comment here to explain the difference of 2 cases

1) In what case, rrefId == forkId
2) why, in this case, don't need to add a fork

File: torch/csrc/distributed/rpc/torchscript_functions.cpp [hunk 2]
```diff
@@ -72,25 +70,61 @@ std::shared_ptr<UserRRef> remoteTorchscript(
...
+    // necessary to introduce python dependency in this file.
+    fm->addCallback([](const Message& message,
+                       const c10::optional<utils::FutureError>& futErr) {
+      RRefContext::handleException(futErr);
+      auto rr = RemoteRet::fromMessage(message);
+      TORCH_INTERNAL_ASSERT(
+          rr->rrefId() == rr->forkId(),
+          "Expecting an OwnerRRef as RemoteRet but got a fork.");
+      auto& refCtx = RRefContext::getInstance();
+      refCtx.delForkOfOwner(rr->rrefId(), rr->rrefId());
```
Comment (xush6528) [inline torch/csrc/distributed/rpc/torchscript_functions.cpp]: I prefer to refactor `finishCreatingOwnerRRef()` into 
`finishCreatingOwnerRRef()` (in rref_ccontetx.cpp callback namespace) and `finishCreatingOwnerRRefWithGIL()` (in python_functions) to reduce duplicate code.

Comment (zhaojuanmao) [reply torch/csrc/distributed/rpc/torchscript_functions.cpp]: future call back schema is defined as void function(const message&, FutureError&), finishCreatingOwnerRRefWithGIL() can not call finishCreatingOwnerRRef(), as it needs finishCreatingOwnerRRef() return deletedRRef?

Comment (xush6528) [reply torch/csrc/distributed/rpc/torchscript_functions.cpp]: Maybe this way?

std::shared_ptr<RRef> finishCreatingOwnerRRef() {

}

For Python remote call
```
    fm->addCallback([](const Message& message,
                       const c10::optional<utils::FutureError>& futErr) {
      auto deletedRRef = callback::finishCreatingOwnerRRef(message, futErr);
     if (deletedRRef && deletedRRef->isPyObj()) {
          AutoGil gil;
          deletedRRef.reset()
     }
```

For Script remote call
```
    fm->addCallback([](const Message& message,
                       const c10::optional<utils::FutureError>& futErr) {
      callback::finishCreatingOwnerRRef(message, futErr);
```

Comment (zhaojuanmao) [reply torch/csrc/distributed/rpc/torchscript_functions.cpp]: sounds good!

File: torch/csrc/distributed/rpc/torchscript_functions.cpp [hunk 3]
```diff
@@ -72,25 +70,55 @@ std::shared_ptr<UserRRef> remoteTorchscript(
...
+        std::move(*scriptRemoteCall).toMessage(),
+        true /*forceGradRecording*/,
+        nullptr);
 
-  return userRRefPtr;
+    // The callback is the same as finishCreatingOwnerRRef() in
+    // python_functions.py, the difference is that ownerRRefPtr created here
+    // always holds an IValue, no need to explicitly hold GIL and clear
+    // deletedRRef. Not calling finishCreatingOwnerRRef() because it is not
+    // necessary to introduce python dependency in this file.
```
Comment (xush6528) [inline torch/csrc/distributed/rpc/torchscript_functions.cpp]: Need to update comment?

**Comment by xush6528** [comment]
Looks good to me. @mrshenli Any more comment?

File: torch/csrc/distributed/rpc/torchscript_functions.cpp [hunk 4]
```diff
@@ -72,25 +70,50 @@ std::shared_ptr<UserRRef> remoteTorchscript(
...
+        true /*forceGradRecording*/,
+        nullptr);
+
+    ctx.addPendingUser(userRRefPtr->forkId(), userRRefPtr);
+    fm->addCallback(callback::confirmPendingUser);
+
+    return userRRefPtr;
+  } else {
+    auto ownerRRefPtr = ctx.createOwnerRRef(returnType);
+    // prevent this owner RRef be deleted due to other forks
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/torchscript_functions.cpp]: be -> from being

File: torch/csrc/distributed/rpc/torchscript_functions.cpp [hunk 5]
```diff
@@ -72,25 +70,50 @@ std::shared_ptr<UserRRef> remoteTorchscript(
...
-  fm->addCallback(callback::confirmPendingUser);
+    auto fm = torch::distributed::autograd::sendMessageWithAutograd(
+        *rpcAgentPtr,
+        dstWorkerInfo,
+        std::move(*scriptRemoteCall).toMessage(),
+        true /*forceGradRecording*/,
+        nullptr);
 
-  return userRRefPtr;
+    fm->addCallback([](const Message& message,
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/torchscript_functions.cpp]: can this be 

```c++
fm->addCallback(callback::finishCreatingOwnerRRef)?
```

**Review by mrshenli — COMMENTED** [review]
The changes look good to me, but there are a lot of tests with irrelevant failures. Shall we rebase and clear the test to get meaningful signals? Thanks!

**Comment by zhaojuanmao** [comment]
test failures are not related

**Review by mrshenli — APPROVED** [review]
Thanks for working on this! Stamp to unblock.

Lint and test failures are irrelevant.

**Pull request closed without merging**